### PR TITLE
Handle metric data type from Akeneo

### PR DESCRIPTION
**Pull request by lruozzi9**
This service handles the metrical attributes in AttributeValueHandler from Akeneo to cast it as string and store it in TEXT Sylius field.

### CODE REVIEW

**Comment by aleho** [comment]
This is a very good addition, in my opinion (and fixes bugs)!

Would adding a mapping for units in `MetricAttributeHandler` make sense to you?
I'm getting e.g. `KILOGRAM` as Akeneo unit, which honestly nobody would use like that in a shop, ever.

I'm thinking of a default mapping array:

```php
const DEFAULT_MAPPINGS = [
    'INCH' => '"', // or maybe ' in' ?
    'KILOGRAM' => ' kg',
   …
];
```

This would also take care of local preferences in display. German Standards for example recommend (mandate?) a space between number and unit, unless it's a single superscript sign (like `"` for inch).

**Comment by lruozzi9** [comment]
Thank you for your contribution @aleho! Actually it would not be useful to show the UNIT code as plain text instead of the symbol for example. On the other hand it would not be convenient and efficient to insert a mapping array within the class itself. Rather we thought of using the Symfony translator to customize the disposition and the unit of measure for specific locale code.

**Comment by aleho** [comment]
The mapping array was my quick and dirty workaround (until I had a look at Akeneo and found that huge list of units it supports).

Are SI units actually translatable? I thought they were supposed to be universal?

**Comment by lruozzi9** [comment]
> Are SI units actually translatable? I thought they were supposed to be universal?

Well, technically yes. But, for example, in Italy we usually don't say 23 in but 23".

**Comment by aleho** [comment]
That's a "standardized" rule, as far as I'm aware of: Single, superscript units - like ° or " - should be used without a space, every other SI unit with either a space or a non-breaking space after the value.

File: spec/Converter/ValueConverterSpec.php [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+        $this->beConstructedWith();
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType(ValueConverter::class);
+    }
+
+    function it_implements_value_handler_interface()
```
Comment (mmenozzi) [inline spec/Converter/ValueConverterSpec.php:77]: it_implements_value_converter_interface

File: spec/Converter/ValueConverterSpec.php [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+    function it_converts_text_value_from_akeneo_to_text_value(
+        AttributeInterface $textAttribute
+    ) {
+        $value = 'Agape';
+        $this->convert($textAttribute, $value, self::IT_LOCALE_CODE)->shouldReturn($value);
+    }
+
+    function it_converts_integer_value_from_akeneo_to_integer_value(
+        AttributeInterface $integerValue
```
Comment (mmenozzi) [inline spec/Converter/ValueConverterSpec.php:147]: $integerAttribute

File: spec/Converter/ValueConverterSpec.php [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    ) {
+        $value = 123;
+        $this->convert($integerValue, $value, self::IT_LOCALE_CODE)->shouldReturn($value);
+    }
+
+    function it_converts_textarea_value_from_akeneo_to_textarea_value(
+        AttributeInterface $textareaValue
```
Comment (mmenozzi) [inline spec/Converter/ValueConverterSpec.php:154]: $textareaAttribute

File: spec/Converter/ValueConverterSpec.php [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    public function let(
+        TranslatorInterface $translator,
+        AttributeInterface $textAttribute,
+        AttributeInterface $checkboxAttribute,
+        AttributeInterface $textareaAttribute,
+        AttributeInterface $integerAttribute,
+        AttributeInterface $selectAttribute,
+        AttributeInterface $datetimeAttribute
+    ) {
+        $translator->trans('webgriffe_sylius_akeneo.ui.value_converter', ['unit' => 'INCH', 'amount' => '23'], null, 'it')->willReturn('23"');
```
Comment (mmenozzi) [inline spec/Converter/ValueConverterSpec.php:49]: I'd rename the translate key to `webgriffe_sylius_akeneo.ui.metric_amount_unit`

File: spec/Converter/ValueConverterSpec.php [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            $textAttribute,
+            [
+                'amount' => 23,
+                'key' => 'value',
+            ],
+            'it',
+        ]);
+    }
+
+    function it_converts_metric_value_from_akeneo_to_text_value_not_translated_when_translator_is_not_injected(
```
Comment (mmenozzi) [inline spec/Converter/ValueConverterSpec.php:111]: Why not always requiring the translator?

Comment (lruozzi9) [reply spec/Converter/ValueConverterSpec.php:111]: To avoid generating a BC break in the AttributeValueHandler definition, we must allow the static implementation of the ValueConverter. To avoid injecting a secondary temporary service (the translator) we must allow the construction without this service. An alternative could be instantiate the service statically like this:
`$valueConverter = new ValueConverter(new IdentityTranslator());`

Comment (mmenozzi) [reply spec/Converter/ValueConverterSpec.php:111]: You're right.

**Review by mmenozzi — APPROVED** [review]

File: src/Converter/ValueConverter.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                throw new \InvalidArgumentException(
+                    sprintf(
+                        'This select attribute can only save existing attribute options. ' .
+                        'Attribute option codes [%s] do not exist.',
+                        implode(', ', $invalid)
+                    )
+                );
+            }
+
+            return [$value];
```
Comment (aleho) [inline src/Converter/ValueConverter.php:68]: Does this work for you? For me it breaks choice selection because the value has already been casted to an array.

(See https://github.com/webgriffe/SyliusAkeneoPlugin/blob/aa1406faa3d281f15594db8d1df09988862ef0ef/src/ValueHandler/AttributeValueHandler.php#L186, which is virtually identical)

Comment (mmenozzi) [reply src/Converter/ValueConverter.php:68]: I think @aleho is right here. @lruozzi9 what do you think?

**Review by mmenozzi — CHANGES_REQUESTED** [review]

File: src/Converter/ValueConverter.php (hunk 6 above)
Comment (lruozzi9) [reply src/Converter/ValueConverter.php:68]: I check right away!

Comment (lruozzi9) [reply src/Converter/ValueConverter.php]: Hi @aleho. You were right. There was an error in a spec test that involved returning an array of arrays for multiple select options. It should now be correct. If you have already installed this PR I ask you to prove that the conversion is now correctly performed. Thanks again for your contribution. 😁

File: README.md [hunk 7]
```diff
@@ -94,7 +94,7 @@ Override Sylius template by create a new file in the folder: `templates/bundles/
 
         # ...
 
-        <a class="ui labeled icon button green" href="{{  path('webgriffe_sylius_akeneo_product_enqueue', {'productId': product.id }) }}" >  
+        <a class="ui labeled icon button violet" href="{{  path('webgriffe_sylius_akeneo_product_enqueue', {'productId': product.id }) }}" >
```
Comment (mmenozzi) [inline README.md:97]: This is not related to this PR. Moreover I think we used `purple` and not `violet` before.

Comment (lruozzi9) [reply README.md]: Yes, I will do it in master directly or open a proper issue. The class is violet not purple, see [here](https://github.com/webgriffe/SyliusAkeneoPlugin/blob/aa1406faa3d281f15594db8d1df09988862ef0ef/src/Resources/config/config.yaml#L66).

File: src/Converter/ValueConverter.php (hunk 6 above)
Comment (aleho) [reply src/Converter/ValueConverter.php]: Yes, that will work. I already fixed it locally by removing the return statement and importing a few hundred products yesterday. Thanks for your work!

**Review by mmenozzi — APPROVED** [review]
